test/jdk/java/lang/invoke/MethodHandleProxies/ProxiesImplementationTest.java

```diff
@@ -0,0 +1,208 @@
+                "Leaked lookup original flag");
+        assertThrows(IllegalAccessException.class, () -> MethodHandles.privateLookupIn(instanceClass,
+                Client.leakLookup()));
+    }
+
+    /**
+     * Tests the Proxy module properties for Proxies implementing system and
+     * user interfaces.
+     */
+    @ParameterizedTest
+    @ValueSource(classes = {Client.class, Runnable.class})
+    public void testModule(Class<?> ifaceClass) {
+        var mh = MethodHandles.zero(void.class);
+
+        var inst = asInterfaceInstance(ifaceClass, mh);
+        Module ifaceModule = ifaceClass.getModule();
+        Class<?> implClass = inst.getClass();
+        Module implModule = implClass.getModule();
+
+        String implPackage = implClass.getPackageName();
+        assertFalse(implModule.isExported(implPackage),
+                "implementation should not be exported");
+        assertTrue(ifaceModule.isExported(ifaceClass.getPackageName(), implModule),
+                "interface package should be exported to implementation");
+        assertTrue(implModule.isOpen(implPackage, MethodHandleProxies.class.getModule()),
+                "implementation class is not reflectively open to MHP class");
+        assertTrue(implModule.isNamed(), "dynamic module must be named");
+        assertTrue(implModule.getName().startsWith("jdk.MHProxy"),
+                () -> "incorrect dynamic module name: " + implModule.getName());
+
+        assertSame(implClass.getClassLoader(), implModule.getClassLoader(),
+                "module class loader should be proxy class's loader");
+    }
+
+    /**
+     * Tests the access control of Proxies implementing system and user
+     * interfaces.
+     */
+    @ParameterizedTest
+    @ValueSource(classes = {Client.class, Runnable.class})
+    public void testNoInstantiation(Class<?> ifaceClass) throws ReflectiveOperationException {
+        var mh = MethodHandles.zero(void.class);
+        var instanceClass = asInterfaceInstance(ifaceClass, mh).getClass();
+        var ctor = instanceClass.getDeclaredConstructor(MethodHandles.Lookup.class, MethodHandle.class, MethodHandle.class);
+
+        assertThrows(IllegalAccessException.class, () -> ctor.newInstance(Client.leakLookup(), mh, mh));
+        assertThrows(IllegalAccessException.class, () -> ctor.newInstance(MethodHandles.lookup(), mh, mh));
+        assertThrows(IllegalAccessException.class, () -> ctor.newInstance(MethodHandles.publicLookup(), mh, mh));
+    }
+
+    /**
+     * Tests the caching and weak reference of implementation classes.
+     */
+    @ParameterizedTest
+    @ValueSource(classes = {Runnable.class, Client.class})
+    public void testWeakImplClass(Class<?> ifaceClass) {
+        var mh = MethodHandles.zero(void.class);
+        WeakReference<Class<?>> cl;
+
+        var c1 = asInterfaceInstance(ifaceClass, mh);
+        cl = new WeakReference<>(c1.getClass());
```

Comment:
Nit: define a new variable not to mix with `c1`.

```suggestion
        var wr = new WeakReference<>(c1.getClass());
```